### ISSUE TITLE
Conflicting approaches to supporting shared properties

### ISSUE THREAD

**Issue by tesujimath** [lumihq/purescript-react-basic#92]
So, I just noticed the new release, which includes #91.  Which addresses a problem which was already addressed by #84.

Maintaining some properties independently of the codegen seems a bit of a hack to me.  Isn't the right way to do this to merge #84 and rollback on the alternate approach?

Note that #84 also addresses #88

**Comment by spicydonuts** [comment]
Yes, I agree. Sorry I thought your PR was still in progress. I'll take another look at it soon!

**Comment by tesujimath** [comment]
@spicydonuts @justinwoo I have reviewed these conflicts, as follows:

| property                         | SharedProps    | codegen                |
|----------------------------------+----------------+------------------------|
| `allowTransparency`              | `String`       | `Boolean`              |
| `autoComplete`                   | `String`       | `Boolean`              |
| `autoFocus`                      | `String`       | `Boolean`              |
| `colSpan`                        | `Number`       | `Int`                  |
| `contentEditable`                | `String`       | `Boolean`              |
| `dangerouslySetInnerHTML`        |                | `{ __html :: String }` |
| `draggable`                      | `String`       | `Boolean`              |
| `formNoValidate`                 | `String`       | `Boolean`              |
| `key`                            | `String`       |                        |
| `maxLength`                      | `String`       | `Int`                  |
| `minLength`                      | `String`       | `Int`                  |
| `noValidate`                     | `String`       | `Boolean`              |
| `onClick`                        | `EventHandler` |                        |
| `onMouseDown`                    | `EventHandler` |                        |
| `onMouseEnter`                   | `EventHandler` |                        |
| `onMouseLeave`                   | `EventHandler` |                        |
| `onMouseMove`                    | `EventHandler` |                        |
| `onMouseOut`                     | `EventHandler` |                        |
| `onMouseOver`                    | `EventHandler` |                        |
| `onMouseUp`                      | `EventHandler` |                        |
| `rowSpan`                        | `Number`       | `Int`                  |
| `spellCheck`                     | `String`       | `Boolean`              |
| `srcDoc`                         | `String`       | `JSX`                  |
| `style`                          | `CSS`          | `String`               |
| `suppressContentEditableWarning` |                | `Boolean`              |
| `tabIndex`                       | `String`       | `Int`                  |
| `unselectable`                   | `String`       | `Boolean`              |
|----------------------------------+----------------+------------------------|

I propose that SharedProps be dropped, in favour of codegen from [react-html-attributes](https://github.com/jackyho112/react-html-attributes)  There are a few issues to be addressed first:

`key` is missing from react-html-attributes, probably should be added

`onClick` is listed in react-html-attributes as an attribute of `a`, but applies more widely, i.e. an attribute of `*`

`onMouse*` events are missing from react-html-attributes, and there are many more:
https://developer.mozilla.org/en-US/docs/Web/API/GlobalEventHandlers

`style` should be added to the codegen consts `export.types` with type `CSS`.

**Comment by spicydonuts** [comment]
I wonder if `key` and `ref` could be made optional extra props by the `element` function itself.. I don't think the Union trick would work there though.

**Comment by tesujimath** [comment]
I think handling key and ref explicitly like you suggest is the way to go.  That's easy.

Regarding different types for different elements, that's already addressed by my codegen PR, https://github.com/lumihq/purescript-react-basic/pull/84#issuecomment-473137352, and see the `typesByElement` in `codegen/consts.js`
